crates/turbopack-ecmascript/src/tree_shake/analyzer/mod.rs

```diff
@@ -0,0 +1,487 @@
+use fxhash::{FxBuildHasher, FxHashMap, FxHashSet};
+use indexmap::IndexSet;
+use petgraph::graphmap::DiGraphMap;
+use swc_core::{
+    common::collections::AHashSet,
+    ecma::{
+        ast::{
+            op, ArrowExpr, AssignExpr, AssignPatProp, Callee, ClassDecl, ClassExpr, Decl,
+            ExportDecl, ExportNamedSpecifier, Expr, FnDecl, FnExpr, Function, Id, Ident,
+            JSXElementName, JSXObject, ModuleExportName, Pat, Program, Prop, VarDeclarator,
+        },
+        atoms::js_word,
+        utils::{collect_decls, find_pat_ids},
+        visit::{noop_visit_type, Visit, VisitWith},
+    },
+};
+
+#[cfg(test)]
+mod tests;
+
+#[derive(Debug, Default)]
+pub struct ItemDepGraph {
+    used_names: FxHashMap<Id, VarInfo>,
+
+    /// Variable usage graph
+    ///
+    /// We use `u32` because [DiGraphMap] stores types as `(N, 1 bit)` so if
+    /// we use u32 it fits into the cache line of cpu.
+    graph: DiGraphMap<u32, VarInfo>,
+    /// Entry points (exports).
+    entries: FxHashSet<u32>,
+
+    graph_ix: IndexSet<Id, FxBuildHasher>,
+}
+
+impl ItemDepGraph {
+    fn node(&mut self, id: &Id) -> u32 {
+        self.graph_ix.get_index_of(id).unwrap_or_else(|| {
+            let ix = self.graph_ix.len();
+            self.graph_ix.insert_full(id.clone());
+            ix
+        }) as _
+    }
+
+    /// Add an edge to dependency graph
+    fn add_dep_edge(&mut self, from: Id, to: Id, live: bool) {
+        let from = self.node(&from);
+        let to = self.node(&to);
+
+        match self.graph.edge_weight_mut(from, to) {
+            Some(info) => {
+                if live {
+                    info.live += 1;
+                } else {
+                    info.initial += 1;
+                }
+            }
+            None => {
+                self.graph.add_edge(
+                    from,
+                    to,
+                    VarInfo {
+                        initial: u32::from(!live),
+                        live: u32::from(live),
+                    },
+                );
+            }
+        };
+    }
+}
+
+#[derive(Debug, Default)]
+struct VarInfo {
+    pub live: u32,
+    pub initial: u32,
+}
+
+pub fn analyze(p: &Program) -> ItemDepGraph {
+    let mut graph = ItemDepGraph::default();
+
+    let mut analyzer = Analyzer {
+        data: &mut graph,
+        scope: Scope::default(),
+        in_var_decl: false,
+        cur_class_id: None,
+        cur_fn_id: None,
+        is_expr_not_live_dep: false,
+    };
+    p.visit_with(&mut analyzer);
```

Comment:
Instead of calling the visitor on the whole program (and needing to visit all declarations, module items, etc)
you could also consider to loop over all top level statements/declarations and call the visitor only for children of them. So you don't need to visit these top level constructs and the visitor could concentrate on visiting identifiers.